### PR TITLE
Redo of Index/Filter/Data blocks sizes in Block (LRU) Block Cache per CF after rebase on RocksDB 8.1 (#516)

### PR DESCRIPTION
**Pull request by udi-speedb**
The original issue is: https://github.com/speedb-io/speedb/issues/338

### CODE REVIEW

**Comment by Yuval-Ariel** [comment]
plz also update the HISTORY.md file

**Comment by udi-speedb** [comment]
> plz also update the HISTORY.md file

OK.

File: db/column_family.cc [hunk 1]
```diff
@@ -662,6 +667,12 @@ ColumnFamilyData::~ColumnFamilyData() {
   prev->next_ = next;
   next->prev_ = prev;
 
+  const BlockBasedTableOptions* bbto =
+      ioptions_.table_factory->GetOptions<BlockBasedTableOptions>();
```
Comment (Yuval-Ariel) [inline db/column_family.cc:671]: is bbto guaranteed to be nullptr if table_factory type is NOT BlockBasedTableOptions?

File: db/internal_stats.h [hunk 2]
```diff
@@ -479,6 +479,13 @@ class InternalStats {
     uint64_t last_start_time_micros_ = 0;
     uint64_t last_end_time_micros_ = 0;
 
+    // Instances of this class are created per cf, but cache stats collection
```
Comment (Yuval-Ariel) [inline db/internal_stats.h:482]: from what i can tell, this is not true. the collector and this (CacheEntryRoleStats) is a shared obj among all cfs

File: db/column_family.cc (hunk 1 above)
Comment (mrambacher) [reply db/column_family.cc:671]: > is bbto guaranteed to be nullptr if table_factory type is NOT BlockBasedTableFactory?

Yes, GetOptions will return null if the table_factory does not contain a BlockBasedTableOptions

**Comment by udi-speedb** [comment]
@Yuval-Ariel - Are you done with the review?

File: db/column_family.cc (hunk 1 above)
Comment (Yuval-Ariel) [reply db/column_family.cc:671]: thanks!

File: db/internal_stats.cc [hunk 3]
```diff
@@ -744,6 +760,38 @@ std::string InternalStats::CacheEntryRoleStats::ToString(
...
 }
 
+std::string InternalStats::CacheEntryRoleStats::CacheOwnerStatsToString(
+    const std::string& cf_name, Cache::ItemOwnerId cache_owner_id) {
+  std::ostringstream str;
+
+  const auto& cf_charges_per_role_pos =
+      charge_per_item_owner.find(cache_owner_id);
+
+  std::vector<CacheEntryRole> roles{CacheEntryRole::kDataBlock,
```
Comment (Yuval-Ariel) [inline db/internal_stats.cc:770]: not necessarily for this part of the code, but obvious here:
why not support all the CacheEntryRole?

File: db/internal_stats.cc [hunk 4]
```diff
@@ -744,6 +760,38 @@ std::string InternalStats::CacheEntryRoleStats::ToString(
...
+                                    CacheEntryRole::kFilterBlock,
+                                    CacheEntryRole::kIndexBlock};
+  std::array<size_t, kNumCacheEntryRoles> roles_total_charge{};
```
Comment (Yuval-Ariel) [inline db/internal_stats.cc:773]: why allocate is size of array kNumCacheEntryRoles when only 3 roles are indexed?

File: db/internal_stats.cc [hunk 5]
```diff
@@ -744,6 +760,38 @@ std::string InternalStats::CacheEntryRoleStats::ToString(
...
+
+  str << "Block cache [" << cf_name << "] ";
+
+  if (cf_charges_per_role_pos != charge_per_item_owner.end()) {
+    const std::array<size_t, kNumCacheEntryRoles>& cf_charges_per_role =
+        cf_charges_per_role_pos->second;
+    for (auto role : roles) {
+      auto role_idx = static_cast<unsigned int>(role);
+      roles_total_charge[role_idx] = cf_charges_per_role[role_idx];
```
Comment (Yuval-Ariel) [inline db/internal_stats.cc:782]: why are you copying the charges to a new container when i think you can pass the values directly from cf_charges_per_role?

File: db/internal_stats.cc (hunk 3 above)
Comment (udi-speedb) [reply db/internal_stats.cc:770]: Block cache entry stats are also displayed for these 3 + wbm. All the rest are misc and are negligible. I do not want to have a long and unreadable log line. In addition, WBM is not known per cf.

File: db/internal_stats.h (hunk 2 above)
Comment (udi-speedb) [reply db/internal_stats.h:482]: OK.

**Comment by udi-speedb** [comment]
> Member

Done

File: db/internal_stats.cc (hunk 4 above)
Comment (udi-speedb) [reply db/internal_stats.cc:773]: std::array must have a compile time size, so, using roles.size() will not compile. As I have removed the array anyway, it's no longer relevant.

File: db/internal_stats.cc (hunk 5 above)
Comment (udi-speedb) [reply db/internal_stats.cc:782]: I have updated the code. No more copying.

File: HISTORY.md [hunk 6]
```diff
@@ -5,6 +5,7 @@
 ### New Features
 * Snapshot optimization - The most important information inside a snapshot is its Sequence number, which allows the compaction to know if the key-value should be deleted or not. The sequence number is being changed when modification happens in the db. This feature allows the db to take a snapshot without acquiring db mutex when the last snapshot has the same sequence number as a new one. In transactional db with mostly read operations, it should improve performance when used with multithreaded environment and as well other scenarios of taking large amount of snapshots with mostly read operations.
 * Add a TablePinningPolicy to the BlockBasedTableOptions.  This class controls when blocks should be pinned in memory for a block based table.  The default behavior uses the MetadataCacheOptions to control pinning and behaves identical to the previous releases. 
+* Redo of Index/Filter/Data blocks sizes in Block (LRU) Block Cache per CF after rebase on RocksDB 8.1
```
Comment (Yuval-Ariel) [inline HISTORY.md:8]: in this release, this is the first reference to this feature, so i think more information is required here such as what exactly the user will gain from it and its limitations. 
also both the issue and PR dont have much information. maybe connect the original issue?

Comment (udi-speedb) [reply HISTORY.md:8]: The HISTORY.md is not the place to document a feature. I have updated the commit message with more detailed information.

Comment (udi-speedb) [reply HISTORY.md]: Done

**Review by Yuval-Ariel — APPROVED** [review]

**Comment by udi-speedb** [comment]
Rebased on latest origin/main and squashed to prepare for the merge

**Review by Yuval-Ariel — APPROVED** [review]

**Comment by Yuval-Ariel** [comment]
checking licence and history

**Pull request merged**